[new module] Add Azure Container Instance support - cloud/azure/azure_rm_containerinstance

zikalino opened the pull request:
##### SUMMARY
Adding Azure Container Instance support.

##### ISSUE TYPE
 - New Module Pull Request

##### COMPONENT NAME
azure_rm_aci

##### ANSIBLE VERSION
```
ansible 2.5.0 (devel 125c4b135d) last updated 2017/10/17 07:30:02 (GMT +000)
  config file = /etc/ansible/ansible.cfg
  configured module search path = [u'/root/.ansible/plugins/modules', u'/usr/share/ansible/plugins/modules']
  ansible python module location = /ansible/lib/ansible
  executable location = /ansible/bin/ansible
  python version = 2.7.12 (default, Nov 19 2016, 06:48:10) [GCC 5.4.0 20160609]
```


##### ADDITIONAL INFORMATION
<!---
Include additional information to help people understand the change here.
For bugs that don't have a linked bug report, a step-by-step reproduction
of the problem is helpful.
  -->

<!--- Paste verbatim command output below, e.g. before and after your change -->
```

```

Commit:
first dirty container instance

Commit:
first dirty container instance

Commit:
first dirty container instance

Commit:
first dirty container instance

Commit:
added my name ;-)

Commit:
added my name ;-)

Commit:
more updates

Commit:
more updates

Commit:
more updates

Commit:
more updates

Commit:
removed unnecessary stuff

Commit:
removed unnecessary stuff

Commit:
container instance updates

Commit:
container instance updates

Commit:
container instance updates

Commit:
container instance updates

Commit:
several fixes

Commit:
several fixes

Commit:
several fixes

Commit:
several fixes

Commit:
undo changes in common

Commit:
undo changes in common

Commit:
undo changes in common

Commit:
undo changes in common

Commit:
removed unnecessary references, fixed delete

Commit:
removed unnecessary references, fixed delete

Commit:
added / updated parameters

Commit:
added / updated parameters

Commit:
added / updated parameters

Commit:
added / updated parameters

Commit:
updated samples & comments

Commit:
updated samples & comments

Commit:
updated docs, comments, samples and added registry credentials

Commit:
updated docs, comments, samples and added registry credentials

Commit:
added ip address and port

Commit:
added ip address and port

Commit:
query existing container instance (but result not used yet)

Commit:
query existing container instance (but result not used yet)

Commit:
query existing container instance (but result not used yet)

Commit:
query existing container instance (but result not used yet)

Commit:
some major changes to the module

Commit:
some major changes to the module

Commit:
more fixes

Commit:
more fixes

Commit:
added requirement for containerinstance module

Commit:
adding integration test

Commit:
adding integration test

Commit:
fixes for pull request

Commit:
fixes for pull request

Commit:
updated version

Commit:
updated version

Commit:
updated version to 2.5

Commit:
updated version to 2.5

Commit:
updated version to 2.5

Commit:
updated version to 2.5

Commit:
updated version

Commit:
updated version

Commit:
updated version

Commit:
updated version

Commit:
updated integration.cloud.azure.txt as requested by test framework

Commit:
updated integration.cloud.azure.txt as requested by test framework

Commit:
removed due to merge reasons

Commit:
removed due to merge reasons

Commit:
removed due to merge reasons

Commit:
removed due to merge reasons

Commit:
updated requirements-azure.txt

Commit:
updated requirements-azure.txt

Commit:
updated requirements-azure.txt

Commit:
updated requirements-azure.txt

Commit:
undone azure-rm-common

Commit:
undone azure-rm-common

Commit:
undone azure-rm-common

Commit:
undone azure-rm-common

Commit:
lf

Commit:
lf

Commit:
properly update test requirements

Commit:
properly update test requirements

Commit:
properly update test requirements

Commit:
properly update test requirements

yuwzho commented on lib/ansible/modules/cloud/azure/azure_rm_aci.py:
```diff
@@ -0,0 +1,503 @@
+            - public
+        required: false
+        default: None
+    port:
```
`ports`, and it should be a list. Actually, I wonder why we need this field, it should be a summary of the container's ports

zikalino replied:
yes, indeed, we could just try to collect all the ports exposed by all containers and make them public, but then maybe somebody just doesn't want all the ports to be public, some may be used internally i imagine. i will change port -> ports though as it's a list.

yuwzho commented on lib/ansible/modules/cloud/azure/azure_rm_aci.py:
```diff
@@ -0,0 +1,503 @@
+        '''
+        self.log("Creating / Updating the ACI instance {0}".format(self.name))
+
+        credentials = None
```
`registryCredentials`? Or it is confused it is registry credential or azure credential

yuwzho commented on lib/ansible/modules/cloud/azure/azure_rm_aci.py:
```diff
@@ -0,0 +1,503 @@
+
+        if self.ip_address is not None:
+            # get list of ports
+            if self.ports:
```
If `self.ports` is nullOrEmpty, but the `self.ip_address` is not null, the `ip_address` should be created with default port 80

zikalino replied:
if self.ip_address is set to public i will make port 80 default

nitzmahone replied:
Rather than an arbitrary default value, I'd vote for either making `ports` required (and empty list acceptable) when IP address is specified, or just leaving the default an empty list (eg, for "outbound-only" container apps).

yuwzho commented on lib/ansible/modules/cloud/azure/azure_rm_aci.py:
```diff
@@ -0,0 +1,503 @@
+            port_list = container_def.get("ports")
+            if port_list:
+                for port in port_list:
+                    ports.append(ContainerPort(port))
```
should open default 80?

zikalino replied:
well, not necessarily, user may not want to expose any port

yuwzho replied:
Consider this scenario, the containers list contains only one container, and both the ports in container and in the parent level is empty, but the ip_address=public is set, should we open the 80 port as default?

nitzmahone replied:
see above

yuwzho commented on lib/ansible/modules/cloud/azure/azure_rm_aci.py:
```diff
@@ -0,0 +1,503 @@
+            containers.append(Container(name, image, ResourceRequirements(ResourceRequests(memory, cpu)), None, ports))
+
+        parameters = ContainerGroup(self.location,
+                                    None,
```
why not pass the `self.tag` to container group

yuwzho commented on lib/ansible/modules/cloud/azure/azure_rm_aci.py:
```diff
@@ -0,0 +1,503 @@
+                memory = 1.5
+            containers.append(Container(name, image, ResourceRequirements(ResourceRequests(memory, cpu)), None, ports))
+
+        parameters = ContainerGroup(self.location,
```
pls add a comment to show the parameter name of `ContainerGroup(self, location, tags=None, containers=None, image_registry_credentials=None, restart_policy=None, ip_address=None, os_type=None, volumes=None)`

xscript replied:
Using named arguments will help readability

yuwzho commented on lib/ansible/modules/cloud/azure/azure_rm_aci.py:
```diff
@@ -0,0 +1,503 @@
+        for container_def in self.containers:
+            name = container_def.get("name")
+            image = container_def.get("image")
+            memory = container_def.get("memory")
```
`memory = container_def.get("memory", 1.5)`, this will assign memory with 1.5 if the dict field is NONE. [ref](https://docs.quantifiedcode.com/python-anti-patterns/correctness/not_using_get_to_return_a_default_value_from_a_dictionary.html)

yuwzho commented on lib/ansible/modules/cloud/azure/azure_rm_aci.py:
```diff
@@ -0,0 +1,503 @@
+                to_be_updated = self.check_need_update(response)
+                if to_be_updated:
+                    self.log('Deleting ACI instance before update')
+                    self.delete_aci()
```
Since the [ip_address](https://github.com/Azure/azure-sdk-for-python/blob/master/azure-mgmt-containerinstance/azure/mgmt/containerinstance/models/ip_address.py#L44) can accept an ip address as parameter, can we note the original ip_address and apply to the new created one for the update issue

zikalino replied:
nope, it's actually read only, can't be specified here

xscript commented on lib/ansible/modules/cloud/azure/azure_rm_aci.py:
```diff
@@ -0,0 +1,503 @@
+                                    self.os_type)
+
+        try:
+            client = ContainerInstanceManagementClient(self.azure_credentials, self.subscription_id)
```
Reuse one client with line#440 and line#457 to avoid create multiple instances.

xscript commented on lib/ansible/modules/cloud/azure/azure_rm_aci.py:
```diff
@@ -0,0 +1,503 @@
+                to_be_updated = self.check_need_update(response)
+                if to_be_updated:
+                    self.log('Deleting ACI instance before update')
+                    self.delete_aci()
```
looks like `--check` will actually delete the existing instance without creating a new one (line#357)?

Commit:
review related updates

Commit:
review related updates

xscript commented on lib/ansible/modules/cloud/azure/azure_rm_aci.py:
```diff
@@ -0,0 +1,500 @@
+        results = dict()
+        to_be_updated = False
+
+        self.mgmt_client = ContainerInstanceManagementClient(self.azure_credentials, self.subscription_id)
```
- We should enable multi-cloud support. The default endpoint is pointing to global Azure.
- You can use `get_mgmt_svc_client` from `AzureRMModuleBase`, which will add Ansible user agent automatically.


You can refer to code snippet at [here](https://github.com/ansible/ansible/blob/devel/lib/ansible/module_utils/azure_rm_common.py#L734-L736).

yaweiw commented on test/integration/targets/azure_rm_aci/tasks/main.yml:
```diff
@@ -0,0 +1,45 @@
+- name: Create sample container group
+  azure_rm_aci:
+    resource_group: "{{ resource_group }}"
+    name: mynewcontainergroup
```
what is the name is already there? maybe name: "aci{{ resource_group | hash('md5') | truncate(7, True, '') }}" better?

zikalino replied:
that's in ansible's test framework

zikalino replied:
ah, yes, i misunderstood.
well, first of all container instance can be created only in an empty resource group, otherwise will fail.
so it's not a problem here.
I don't really know mechanics of ansible test framework, but seems like every subsequent test run gives me empty resource group

yaweiw commented on test/integration/targets/azure_rm_aci/tasks/main.yml:
```diff
@@ -0,0 +1,45 @@
+  assert:
+    that:
+      - output.changed
+      - output.state.provisioning_state == 'Creating'
```
"Creating" or "Succeeded"?

zikalino replied:
At the moment we are not waiting for "Succeeded". We could but that would require additional code, as create_or_update leaves with "Creating" state and it's work is done.

zikalino replied:
just checked what happens if we try to create 2 containers and one of them fails.
in such case it will stay in "Creating" state and keep retrying to start failed container.

yaweiw commented on test/integration/targets/azure_rm_aci/tasks/main.yml:
```diff
@@ -0,0 +1,45 @@
+    resource_group: "{{ resource_group }}"
+    name: mynewcontainergroup
+    os_type: linux
+    ip_address: public
```
self.resource_group, self.name are only necessary for deletion?

Commit:
review related fixes

Commit:
review related fixes

Commit:
minor fix - sanity

Commit:
minor fix - sanity

Commit:
fix one issue after rebasing

Commit:
fix one issue after rebasing

mattclay commented on sh.exe.stackdump:
```diff
@@ -0,0 +1,9 @@
+Stack trace:
```
This file shouldn't be in the PR.

mattclay commented on test/integration/targets/azure_rm_aci/aliases:
```diff
@@ -0,0 +1,2 @@
+cloud/azure
+destructive
```
Add the `posix/ci/cloud/group2/azure` alias so the tests can run as part of CI.

zikalino replied:
Thanks Matt! I wasn't aware it's needed to run CI. Now actually my test is running... and failing...

"Error creating the ACI instance: Azure Error: MissingSubscriptionRegistration\nMessage: The subscription is not registered to use namespace 'Microsoft.ContainerInstance'

I guess this has to be somehow fixed in the subscription used by CI...

So for now i have to disable it until it's enabled.

(the same applies to ACS and ACR)

mattclay replied:
I've registered `Microsoft.ContainerInstance`, so the tests should be able to run now.

yaweiw replied:
@mattclay wanted to check have you also registered container register? Reference: #32481

mattclay replied:
I've registered `Microsoft.ContainerRegistry` as well.

mattclay commented on test/runner/requirements/requirements.cloud.azure.txt:
```diff
@@ -0,0 +1,17 @@
+packaging
```
This file shouldn't be here. The integration test requirements for Azure come from `test/runner/requirements/integration.cloud.azure.txt`.

Commit:
removed files accidentally added while rebasing

Commit:
removed files accidentally added while rebasing

Commit:
removed checking for changes

Commit:
removed checking for changes

Commit:
several fixes

Commit:
several fixes

Commit:
fixed sanity

Commit:
fixed sanity

lmazuel commented on packaging/requirements/requirements-azure.txt:
```diff
@@ -13,4 +13,5 @@ azure-mgmt-keyvault>=0.40.0,<0.41
 azure-mgmt-web>=0.32.0,<0.33
-azure-mgmt-containerservice>=1.0.0
\ No newline at end of file
+azure-mgmt-containerservice>=1.0.0
```
Beware that latest is 2.0.0. You should change this to >=2.0.0,3.0.0 and check you're ok with 2.0.0

zikalino replied:
this PR is not touching containerservice at all, so it's better not to change it here

lmazuel commented on packaging/requirements/requirements-azure.txt:
```diff
@@ -13,4 +13,5 @@ azure-mgmt-keyvault>=0.40.0,<0.41
-azure-mgmt-containerservice>=1.0.0
\ No newline at end of file
+azure-mgmt-containerservice>=1.0.0
+azure-mgmt-containerinstance>=0.1.0
```
Beware that latest is 0.2.0. You should change this to >=0.2.0,0.3.0 and check you're ok with 0.2.0

lmazuel commented on test/runner/requirements/integration.cloud.azure.txt:
```diff
@@ -13,4 +13,5 @@ azure-mgmt-keyvault>=0.40.0,<0.41
 azure-mgmt-web>=0.32.0,<0.33
-azure-mgmt-containerservice>=1.0.0
\ No newline at end of file
+azure-mgmt-containerservice>=1.0.0
```
See my post before o that

Commit:
updates as requested by reviewers

Commit:
updates as requested by reviewers

Commit:
removed ci as it doesn't work

Commit:
removed ci as it doesn't work

yaweiw commented:
shipit

yaweiw reviewed (commented):
@zikalino please test see if it works. thanks mate.

Commit:
reenabled ci

Commit:
reenabled ci

xscript commented:
shipit

nitzmahone commented on lib/ansible/modules/cloud/azure/azure_rm_aci.py:
```diff
@@ -0,0 +1,471 @@
+            - linux
+            - windows
+        default: linux
+        required: false
```
`required: false` is not necessary to state (it's assumed as the default for `required`)

nitzmahone commented on lib/ansible/modules/cloud/azure/azure_rm_aci.py:
```diff
@@ -0,0 +1,471 @@
+
+DOCUMENTATION = '''
+---
+module: azure_rm_aci
```
Might want to consider a little more descriptive name; maybe `azure_rm_containerinstance` or `azure_rm_containerinstance_containergroup`? Suspect we're going to run into acronym hell here pretty quickly if all the modules are named with just initials...

nitzmahone commented on lib/ansible/modules/cloud/azure/azure_rm_aci.py:
```diff
@@ -0,0 +1,471 @@
+            self.log("Response : {0}".format(response))
+            self.log("ACI instance : {0} found".format(response.name))
+        except CloudError as e:
+            self.log('Did not find the ACI instance.')
```
should check `e.error.error == 'ResourceNotFound'` before assuming that was the error

nitzmahone commented on lib/ansible/modules/cloud/azure/azure_rm_aci.py:
```diff
@@ -0,0 +1,471 @@
+                self.results['changed'] = True
+                self.log("ACI instance deleted")
+            elif self.state == 'present':
+                self.log("Need to check if container group has to be deleted or may be updated")
```
Ideally we'd also actually check at least the basic "shape" of the specified containers as part of the change detection (assuming this resource type supports update, which the REST API and SDK appear to). This is always something that could be added later too, but nice to have if the underlying resource really does support changes.

zikalino replied:
actually in case of container instance, it's not possible to update anything. any attempt to update existing container just fails. it's difficult to judge based on passed parameters whether container instance should be updated or not, for example user may want to update container instance because image has changed. well, in such case they could use tags do detect that, but there are a few other things which make it a bit tricky...

nitzmahone commented on lib/ansible/modules/cloud/azure/azure_rm_aci.py:
```diff
@@ -0,0 +1,471 @@
+        description:
+            - The name of the container group.
+        required: true
+        default: null
```
default is kinda mutually exclusive with `required: true` and doesn't need to be specified here

nitzmahone commented on lib/ansible/modules/cloud/azure/azure_rm_aci.py:
```diff
@@ -0,0 +1,471 @@
+        description:
+            - The IP address type of the container group.
+        choices:
+            - public
```
Even though the underlying SDK is kinda strange about this, I'd suggest adding a none value here and making that the default- IMO it's more readable and nice to allow someone to explicitly state "none" rather than just omitting the arg.

nitzmahone commented on lib/ansible/modules/cloud/azure/azure_rm_aci.py:
```diff
@@ -0,0 +1,471 @@
+    ports:
+        description:
+            - List of ports exposed within the container group.
+        required: false
```
+ default: [] (or whatever you guys decide is appropriate- see discussion below)

nitzmahone reviewed (changes requested):
A few minor things as marked, but looks pretty good otherwise! Doesn't support all REST API capabilities yet (eg container envvars/commands, non-TCP ports), but UI as-designed would easily allow for those to be added later without breaking compatibility.

Thanks!

Commit:
renamed container instance, removed required: false

Commit:
renamed container instance, removed required: false

Commit:
removed default: null

Commit:
removed default: null

Commit:
final updates according to the review

Commit:
final updates according to the review

Commit:
one more fix

Commit:
one more fix

Commit:
changed location as default from resource group can't handle containers

zikalino commented:
ready_for_review

zikalino commented:
@nitzmahone could you check again? i think all the comments are fixed now.

zikalino commented:
@nitzmahone could you check again? i think all the comments are fixed now.

zikalino commented:
@nitzmahone  please recheck this module

zikalino commented:
ready_for_review

jborean93 commented:
I haven't gone into a full review of this module but 2 things stand out that would need to be fixed
* The return object should not be `state` with the output from the SDK but unknown values like the ID as their own keys
* The tests should be expanded to cover check mode, idempotency and any update actions that are normally done.

Commit:
updates to container instance

Commit:
fixed mistakes during merge

Commit:
one more fix

Commit:
another mistake

Commit:
container instance fixes

zikalino commented:
ready_for_review

zikalino commented:
@jborean93  @nitzmahone  could you check this one again?

jborean93 commented on lib/ansible/modules/cloud/azure/azure_rm_containerinstance.py:
```diff
@@ -0,0 +1,429 @@
+    location:
+        description:
+            - Valid azure location. Defaults to location of the resource group.
+        default: resource_group location
```
Unless it's a specific value, leave the default out and put it in the description like you have done.

jborean93 commented on lib/ansible/modules/cloud/azure/azure_rm_containerinstance.py:
```diff
@@ -0,0 +1,429 @@
+                    - List of ports exposed within the container group.
+    force_update:
+        description:
+            - Force update of existing container instance.
```
Probably expand on this a bit, maybe say if false then an existing container will not be changed and so on.

jborean93 commented on lib/ansible/modules/cloud/azure/azure_rm_containerinstance.py:
```diff
@@ -0,0 +1,429 @@
+            ),
+            registry_password=dict(
+                type='str',
+                default=None
```
Should set `no_log=True` for fields like this, this stops Ansible from logging secret information to plaintext sources.

jborean93 commented on lib/ansible/modules/cloud/azure/azure_rm_containerinstance.py:
```diff
@@ -0,0 +1,429 @@
+        self.mgmt_client = self.get_mgmt_svc_client(ContainerInstanceManagementClient,
+                                                    base_url=self._cloud_environment.endpoints.resource_manager)
+
+        try:
```
I believe you don't need to put this in a try/except anymore, the logic is done in azure_rm_common.

jborean93 replied:
@zikalino are you able to update this?

jborean93 commented on lib/ansible/modules/cloud/azure/azure_rm_containerinstance.py:
```diff
@@ -0,0 +1,429 @@
+            self.log("Need to Create / Update the container instance")
+
+            if self.check_mode:
+                return self.results
```
You would need to set `changed` if `force_update` is set even in check mode.

jborean93 commented on lib/ansible/modules/cloud/azure/azure_rm_containerinstance.py:
```diff
@@ -0,0 +1,429 @@
+          - 81
+'''
+RETURN = '''
+state:
```
This needs to be updated to reflect the actual return values.

jborean93 commented on lib/ansible/modules/cloud/azure/azure_rm_containerinstance.py:
```diff
@@ -0,0 +1,429 @@
+
+try:
+    from msrestazure.azure_exceptions import CloudError
+    from azure.mgmt.containerinstance.models import (ContainerGroup,
```
Don't we need to pin the modules on a specific version now to ensure API changes don't break the code?

jborean93 replied:
@zikalino Are you able to confirm this, my assumption is that we needed to pin the modules to a specific version.

jborean93 commented on lib/ansible/modules/cloud/azure/azure_rm_containerinstance.py:
```diff
@@ -0,0 +1,429 @@
+                ports = []
+                for port in self.ports:
+                    ports.append(Port(port, "TCP"))
+                ip_address = IpAddress(ports, self.ip_address)
```
Can this be kwargs instead?

jborean93 commented on lib/ansible/modules/cloud/azure/azure_rm_containerinstance.py:
```diff
@@ -0,0 +1,429 @@
+                for port in port_list:
+                    ports.append(ContainerPort(port))
+
+            containers.append(Container(name, image, ResourceRequirements(ResourceRequests(memory, cpu)), None, ports))
```
kwargs?

jborean93 commented on lib/ansible/modules/cloud/azure/azure_rm_containerinstance.py:
```diff
@@ -0,0 +1,429 @@
+
+        super(AzureRMContainerInstance, self).__init__(derived_arg_spec=self.module_arg_spec,
+                                                       supports_check_mode=True,
+                                                       supports_tags=True)
```
I don't see anything in the code that seems to support tags, are you able to set this to False and remove the azure_tags document fragment or add tag support?

jborean93 commented on lib/ansible/modules/cloud/azure/azure_rm_containerinstance.py:
```diff
@@ -0,0 +1,429 @@
+            self.log("Container instance already exists")
+
+            if self.state == 'absent':
+                self.delete_containerinstance()
```
Should not run when in check mode.

jborean93 commented on test/integration/targets/azure_rm_containerinstance/tasks/main.yml:
```diff
@@ -0,0 +1,101 @@
+- name: Create sample container instance
```
Need some more tests with check mode and idempotence runs around removing the instance.

Commit:
several fixes to container instance

Commit:
return value fix

Commit:
minor update

Commit:
just one api version right now

Commit:
fixed api version

Commit:
container instance does not suppurt api version

Commit:
removed unnecessary try blocks

Commit:
removed tags related things

Commit:
fixed pep8

Commit:
final fixes?

zikalino commented:
@jborean93 i think all the comments are resolved now and it's ready to merge.

jborean93 commented on lib/ansible/modules/cloud/azure/azure_rm_containerinstance.py:
```diff
@@ -0,0 +1,429 @@
+ip_address:
+    description:
+        - Public IP Address of created container group.
+    returned: always
```
Is this always returned or only returned when `ip_address: public` is set?

jborean93 commented on lib/ansible/modules/cloud/azure/azure_rm_containerinstance.py:
```diff
@@ -0,0 +1,429 @@
+                self.log("Container instance deleted")
+            elif self.state == 'present':
+                self.log("Need to check if container group has to be deleted or may be updated")
+                to_be_updated = self.force_update
```
While this is just nitpicking, why set this to another variable where you can just check `self.force_update`, they are the same variable.

jborean93 reviewed (changes requested):
Looks good, just a few questions.

Commit:
final updates to the module

Commit:
more fixes

zikalino commented:
ok, fixed these additional minor issues. regarding version. Right now there's only one API version, so there' s no way to pin (that's how SDK works)
In addition i have added this restriction ">=0.2.0,<0.3.0".
And in the next version i am going to get rid of these imports entirely (just like in the new modules).

zikalino commented:
@jborean93 could we merge this?
